# scGET
**s**ingle **c**ell **G**enome and **E**pigenome by **T**ransposases sequencing (**scGET**-seq) aims to discriminate between open accessible chormatin regions and closed compacted chromatin regions within each single cell, taking advantage of two different transposases: transposase-5 binds to *accessible chromatin* (**tn5**) and transposase-H, a chimeric form of tn5 (**tnH**), which binds *compacted chromatin*.


The workflow for analysis of scGET-generated data is based on the *Snakemake*: a workflow management system, which guarantees the possibility to **parallelize** independent jobs. **scGET** workflow is described by the image below: it starts from the *FASTQ* files generated by scGET-seq and it terminates with the production of 2 *BAM* files, accounting for open chromatin regions (**tn5_bam**) and closed chromatin regions (**tnH_bam**).


 ![dag](dag.svg)
# Installing scGET
Before getting your hands dirty with scGET analyses, it is necessary to create a suitable conda environment. However, some of the packages requires different installation procedures. Therefore, we have designed a 3-step process, allowing an easy and quick generation of the *scget environment*.


1. The conda *environment* can be automatically generated, along with the installation of the *majority of required packages* thanks to the **scget.yaml** file:
```
conda env create -f scget.yaml
conda activate scget
```
2. After having activated the *scget* environment, it is necessary to install the **TagDust** package. First, the package must be downloaded and compiled; second, from the *tagdust* directory we can copy the binary tagdust file in our *scget* environment: 
```
tar -zxvf tagdust-2.33.tar.gz 
cd tagdust
./configure 
make
make check
cp ./src/tagdust $CONDA_PREFIX/bin
```
3. The last step accounts for the cloning of **scatACC** repository from github:
```
git clone https://github.com/dawe/scatACC
```
In order to perform the analysis through the calculus cluster, it may be useful to check if `screen` package is already installed:
```
screen --version
```
Output:
>Screen version 4.08.00 (GNU) 05-Feb-20


If `screen` has **not** been **installed** yet, it could be easily installed via `sudo`:
```
sudo apt update
sudo apt install screen
```

 
# Set up
Finally, there are two small tricks left to perform, before you can finally use scGET.


### Cluster set up
 - In your home directory check if you have a snakemake folder inside ``~/.config``. Inside this folder create a slurm folder add in a *config.yaml* file:
```
mkdir -p /home/.config/snakemake/slurm
vi /home/.config/snakemake/slurm/config.yaml
```
 - compile the *config.yaml* file with the content specified below (remember to update the queue name specified by the ``-p`` option and  your ``mail-user``):
```
jobs: 34
cluster: "sbatch --mem={resources.mem_mb} -c {resources.cpus} -o logs_slurm/{SAMPLE_NAME}/{rule}_{wildcards} -e logs_slurm/{SAMPLE_NAME}/{rule}_{wildcards} --mail-type=FAIL --mail-user=user.mail@hsr.it"
default-resources: [cpus=1, mem_mb=2000]
resources: [cpus=10, mem_mb=50000]
use-conda: true
```

### scatACC path
The scGET analysis must **start** from the **scGET directory**. The *scatACC library* is used for one the last step of scGET analysis: therefore, the location of **scatACC directory** must be clarified in order to perform a succesful analysis.


*WHERE SHOULD I CLARIFY THE PATH FOR SCATACC DIRECTORY?*


Let's exemplify! 


Assuming that the **scGET** directory is located in our **home** directory (*/home/scGET*), while our **scatACC** directory is situated in a directory called *repositories* (*/home/repositories/scatACC*):
- We should open the *config.yaml* file present in **scGET** directory:
```
cd /home/scGET
vi config.yaml
```
Output:
>sample: ''
>reads: [1,2,3]
>barcodes: {'tn5':['CGTACTAG','TCCTGAGC','TCATGAGC','CCTGAGAT'],'tnh':['TAAGGCGA','GCTACGCT','AGGCTCCG','CTGCGCAT']}
>genome: /home/scGET_files/genome/hg38.fa
>threads: 8
>cell_number: 5000
>scatacc_path: '/home/scatACC'
>input_path: ''
>input_list: ''
>output_path: ''


- After that we must modify the field *scatacc_path*, specifying our **actual scatACC path**:


Output:
>sample: ''
>reads: [1,2,3]
>barcodes: {'tn5':['CGTACTAG','TCCTGAGC','TCATGAGC','CCTGAGAT'],'tnh':['TAAGGCGA','GCTACGCT','AGGCTCCG','CTGCGCAT']}
>genome: /home/scGET_files/genome/hg38.fa
>threads: 8
>cell_number: 5000
>scatacc_path: '/home/**repositories**/scatACC'
>input_path: ''
>input_list: ''
>output_path: ''

# How to use
standard input 

standard output

for more read advanced.md
